### ISSUE TITLE
Cannot create multiple virtuelenvs on single chef run

### ISSUE THREAD

**Issue by AlexLoup** [poise/poise-python#95]
I am trying to loop through a hash of attributes to set up multiple virtual environments with a single chef recipe using the `python_virtualenv` resource but receive the following error:

```shell
Expected process to exit with [0], but received '1'
           ---- Begin output of ["/root/pythonuser/python_venv/bin/python", "-m", "virtualenv", "/root/otheruser/openstack_venv/"] ----
STDOUT:
STDERR: /root/pythonuser/python_venv/bin/python: No module named virtualenv
           ---- End output of ["/root/pythonuser/python_venv/bin/python", "-m", "virtualenv", "/root/otheruser/python_venv/"] ----
           Ran ["/root/pythonuser/python_venv/bin/python", "-m", "virtualenv", "/root/otheruser/python_venv/"] returned 1
```

It looks like the new pythonuser venv is automatically sourced after it is created and cannot create more venvs since virtualenv is not installed in that environment. Is there a way to `deactivate` the virtual environment after it's created and install other virtualenvs?

This is what the relevant part of the recipe looks like:

```ruby
node['python']['virtenv'].each do |virt, att|
  user node['python']['virtenv'][virt]['user'] do
      uid node['python']['virtenv'][virt]['uid']
      shell '/bin/bash'
      home node['python']['virtenv'][virt]['home']
  end
  template node['python']['virtenv'][virt]['path'] + 'requirements.txt' do
    source 'requirements.txt.erb'
    owner node['python']['virtenv'][virt]['user']
    group 'venv'
    mode node['python']['virtenv'][virt]['file_mode']
    variables(
      packages: node['python']['virtenv'][virt]['requirements']
    )
  end
  python_virtualenv node['python']['virtenv'][virt]['path'] do
    user node['python']['virtenv'][virt]['user']
    group 'venv'
    action :create
  end
  pip_requirements 'requirements.txt' do
    path node['python']['virtenv']['ostadm']['path']
    virtualenv node['python']['virtenv']['ostadm']['path']
  end
end
```

**Comment by coderanger** [comment]
So this is because Poise has a concept of "parent resources". That's how all the various Python resources know which Python executable to use. Both `python_runtime` and `python_virtualenv` count as parents, which is why `pip_requirements` can use the `python_virtualenv` like it does. Your `python_virtualenv` in there isn't specifying a parent so it uses whatever available parent was declared most recently, which the first time through the loop is probably the default `python_runtime` from the `poise-python::default` recipe, but the second time through the loop that's the `python_virtualenv` from the previous loop.

tl;dr set an explicit parent on the `python_virtualenv` like `python '2'` or `python '/usr/bin/python'` (i.e. either a resource name or path to a Python executable.

You could also refactor that loop into a custom resource which would then work a little more like you expect by default :)

**Comment by AlexLoup** [comment]
Ah, that's what I was missing! That solved it, thanks. :)

**Comment by mechmind** [comment]
Why this even works in a such way? My deployment just broke because some other recipe decided to use virtualenv now and packages from target recipe end up in unrelated virtualenv.

**Comment by coderanger** [comment]
@mechmind because it produces the correct UX the majority of the time, which is why you hadn't even noticed it before :)

**Comment by mechmind** [comment]
It should localize it effect to recipe or at least cookbook. Cross-cookbook contamination with hidden state is not "correct UX".

**Comment by dwilches** [comment]
Can this behaviour be localized to a single recipe? Just found an issue similar to @mechmind 's, where someone decided to include a virtualenv in one of the recipes, and then every other recipe that was included later started polluting the first recipe's virtualenv.

In a normal terminal I would issue the `deactivate` command to avoid this problem, can there be a `deactivate` option in the `python_virtualenv` resource?

Thanks.

**Comment by coderanger** [comment]
@dwilches That is what the application cookbook is for, it creates scoped versions of these things.

**Comment by dwilches** [comment]
Ha, thanks a lot!
I didn't know about that cookbook. For others coming here for the same, [this is the link](https://github.com/poise/application_python).
